### PR TITLE
Fee distributor contract

### LINKED ISSUE #4399: Reserve contract receives JS metering fees; IST fees distributed to reward pool and reserve
This ticket covers inflows to the Reserve as well as how the Reserve covers shortfalls from vault liquidations

Inflows to the Reserve should include:
- [x] All IST generated from execution fees (gas) on the Agoric chain should flow to an IST purse controlled by the Reserve #5427
- [x] 50% of fees collected from Inter Protocol contracts go to the Reserve.  Best if this is a governable parameter. #5448

Reward distribution mechanics
- [x] Daily, 10% of total reward purse should be distributed to stakers/validators as rewards (simple smoothing function) #5427

### PR DESCRIPTION
**Pull request by michaelfig**
closes: #4399

## Description

Split fees 50% between rewards and reserve.

### CODE REVIEW

File: packages/governance/test/unitTests/test-committee.js
Comment (turadg) [inline packages/governance/test/unitTests/test-committee.js:77]: why this change?

In my experience, putting lots in the Ava log makes the actual failure scroll off screen. E.g. I just made a change in the other direction: https://github.com/Agoric/agoric-sdk/pull/5440/commits/109145e616704826168799331917082fbff3a3c0

File: packages/deployment/Dockerfile.sdk
Comment (turadg) [inline packages/deployment/Dockerfile.sdk:18]: necessary? (ok if not, just want to know)

File: packages/run-protocol/src/feeDistributor.js
Comment (turadg) [inline packages/run-protocol/src/feeDistributor.js]: please define this on the function declaration instead a callback. that when when you navigated to a part of the type you end up at the implementation as well.

File: packages/run-protocol/src/feeDistributor.js
Comment (turadg) [inline packages/run-protocol/src/feeDistributor.js]: why create a named type for this? I see you want to document `params.epochInterval`. Please do that with https://jsdoc.app/tags-param.html#parameters-with-properties on the function signature.

File: packages/run-protocol/src/feeDistributor.js
Comment (turadg) [inline packages/run-protocol/src/feeDistributor.js]: fully typing this isn't worth it, but `unknown` is always safer than `any`.
```suggestion
 * @param {() => Promise<unknown>} schedulePayments - distribute to the destinations
```

File: packages/run-protocol/src/feeDistributor.js
Comment (turadg) [inline packages/run-protocol/src/feeDistributor.js]: in seconds? if so, should the timer service also be specified as in seconds?

File: packages/run-protocol/src/feeDistributor.js
Comment (turadg) [inline packages/run-protocol/src/feeDistributor.js]: only this function needs to be exported, not the `startDistributing` one.

both of them "start distributing" but over different scopes. the non-exported helper could be `startDistributingPayments` and this one `startDistributingFromCollectors`.

please stop the unnecessary export and make the exported name include "start".

File: packages/run-protocol/src/feeDistributor.js
Comment (turadg) [inline packages/run-protocol/src/feeDistributor.js]:
```suggestion
 * @param {Record<Keyword, ERef<FeeDestination>>} [destinations]
 * @param {Record<Keyword, bigint>} [keywordShares]
```

File: packages/run-protocol/src/feeDistributor.js
Comment (turadg) [inline packages/run-protocol/src/feeDistributor.js:105]: 👍

File: packages/run-protocol/src/feeDistributor.js
Comment (turadg) [inline packages/run-protocol/src/feeDistributor.js]: consider s/total/payment to keep clear what the value is of

File: packages/run-protocol/src/feeDistributor.js
Comment (turadg) [inline packages/run-protocol/src/feeDistributor.js]: doesn't hurt but this is inferred from 167

File: packages/run-protocol/src/feeDistributor.js
Comment (turadg) [inline packages/run-protocol/src/feeDistributor.js]: consider,
```suggestion
      return /** @type {const} */ ([destination, { brand, value }]);
```

File: packages/run-protocol/src/feeDistributor.js
Comment (turadg) [inline packages/run-protocol/src/feeDistributor.js]: does `splitMany` do that assertion? I didn't see that in the docs but I'll take your word :) 

if so, would help to clarify that:
```suggestion
  // Split the payment, which ensures a conservation of the total.
```

File: packages/run-protocol/src/feeDistributor.js
Comment (turadg) [inline packages/run-protocol/src/feeDistributor.js]: please jsdoc the param

File: packages/run-protocol/src/feeDistributor.js
Comment (turadg) [inline packages/run-protocol/src/feeDistributor.js]: please type just the `zcf` param so the return type can be inferred. even better to define the terms like so:
[
```suggestion
/**
 * @param {ZCF} zcf
  */
```

even better to define the terms like so ](https://github.com/Agoric/agoric-sdk/blob/9156f614b7ea8ff8a1862194317d8ebf6324e9a2/packages/run-protocol/src/vaultFactory/liquidateIncrementally.js#L49-L62)

File: packages/run-protocol/src/feeDistributor.js
Comment (turadg) [inline packages/run-protocol/src/feeDistributor.js]: the `...rest` makes the types a little loose. please put it under a keyword `keywordShares` key or explain why that's worse.

File: packages/run-protocol/src/feeDistributor.js
Comment (turadg) [inline packages/run-protocol/src/feeDistributor.js:128]: consider also returning if there are no shares. It's not immediately obvious that this does nothing in that case.

File: packages/run-protocol/src/feeDistributor.js
Comment (turadg) [inline packages/run-protocol/src/feeDistributor.js]: please type facet methods

File: packages/run-protocol/src/feeDistributor.js
Comment (turadg) [inline packages/run-protocol/src/feeDistributor.js]: why?

File: packages/run-protocol/src/feeDistributor.js
Comment (turadg) [inline packages/run-protocol/src/feeDistributor.js]: consider `removeFeeSource` (it will still exist, just removed from this collector)

File: packages/run-protocol/src/feeDistributor.js
Comment (turadg) [inline packages/run-protocol/src/feeDistributor.js]:
```suggestion
     * @param {unknown} target
     * @param {PropertyKey} makeInvitationMethod
     * @param {unknown[]} [args]
```

File: packages/run-protocol/src/feeDistributor.js
Comment (turadg) [inline packages/run-protocol/src/feeDistributor.js]:
```suggestion
/** @typedef {Awaited<ReturnType<typeof start>>['creatorFacet']} FeeDistributorCreatorFacet */
/** @typedef {Awaited<ReturnType<typeof start>>['publicFacet']} FeeDistributorPublicFacet */
```
the reason this didn't error before is the `ReturnType<typeof start>` is `any` due to `ContractStartFn`

**Review by turadg — CHANGES_REQUESTED** [review]
covered what I could before my meeting rn.

File: packages/deployment/Dockerfile.sdk
Comment (michaelfig) [reply packages/deployment/Dockerfile.sdk:18]: Yes, necessary for latest instagoric.  Sorry, should've marked this as a drive-by.

File: packages/governance/test/unitTests/test-committee.js
Comment (michaelfig) [reply packages/governance/test/unitTests/test-committee.js:77]: Console pollution for successful tests really irks me; it makes actual failures scroll off the screen.

File: packages/run-protocol/src/feeDistributor.js
Comment (michaelfig) [reply packages/run-protocol/src/feeDistributor.js]: No, it's in the timer services underlying unit, which may not be seconds (it may be epochs for example).  I'll specify it.

File: packages/governance/test/unitTests/test-committee.js
Comment (michaelfig) [reply packages/governance/test/unitTests/test-committee.js:77]: I think the solution in both of our cases is to buy a bigger scrollback buffer. :)

Comment (michaelfig) [reply packages/governance/test/unitTests/test-committee.js:77]: I agree with your change to gov-collateral, but I find that the manualTimers are not noisy enough to warrant sending them anywhere but `t.log`.

Comment (turadg) [reply packages/governance/test/unitTests/test-committee.js:77]: ah, yeah me too. didn't realize that `t.log` only prints if the test fails.

though because not all logging has access to `t.log` this change makes it so the stream is messages is separated into two different places, making it harder to follow the messages.

what I've done in the past to keep logs clean is override `console` methods with a buffer. we could do something for Ava like that which only prints the console if the test fails. WDYT?

for now humbling requesting that this change be separated to another PR where it can be considered more holistically, but won't block on it.

**Review by turadg — APPROVED** [review]
Some requests weren't addressed but I'll approve to save time.

File: packages/run-protocol/src/feeDistributor.js
Comment (michaelfig) [reply packages/run-protocol/src/feeDistributor.js]: Thanks for the advice.  I extracted the terms with `Parameters<...>`.

File: packages/run-protocol/src/feeDistributor.js
Comment (michaelfig) [reply packages/run-protocol/src/feeDistributor.js]: To both ensure we don't have a bad fee collector, and to catch up in case we're adding a fee collector much later than we started distributing fees.

File: packages/run-protocol/src/feeDistributor.js
Comment (michaelfig) [reply packages/run-protocol/src/feeDistributor.js]: I opted for `startPeriodicCollection` and a `stop` method on its return value.

File: packages/run-protocol/src/feeDistributor.js
Comment (turadg) [inline packages/run-protocol/src/feeDistributor.js:201]: 👌 naming

File: packages/run-protocol/src/feeDistributor.js
Comment (turadg) [inline packages/run-protocol/src/feeDistributor.js:89]: not imported. consider not exporting.

File: packages/run-protocol/src/feeDistributor.js
Comment (turadg) [inline packages/run-protocol/src/feeDistributor.js:114]: not imported. consider not exporting.

File: packages/run-protocol/src/feeDistributor.js
Comment (turadg) [inline packages/run-protocol/src/feeDistributor.js]: not imported. consider not exporting. TMK, contracts are to export only `start`.

also shouldn't this be `const makeContractFeeCollector = (zoe, creatorFacet) => {` ?

File: packages/run-protocol/src/feeDistributor.js
Comment (turadg) [inline packages/run-protocol/src/feeDistributor.js]: same as above, not imported and should be const

File: packages/run-protocol/test/test-feeDistributor.js
Comment (turadg) [inline packages/run-protocol/test/test-feeDistributor.js:12]: I like that you're testing the object. is anything testing the contract?

**Review by turadg — APPROVED** [review]
Outstanding concerns but nothing that should delay security review.

File: packages/run-protocol/test/test-feeDistributor.js
Comment (michaelfig) [reply packages/run-protocol/test/test-feeDistributor.js:12]: > I like that you're testing the object. is anything testing the contract?

Yes, the `run-protocol/test/test-gov-collaterals.js` is testing it.

File: packages/run-protocol/src/feeDistributor.js
Comment (michaelfig) [reply packages/run-protocol/src/feeDistributor.js]: I'm exporting for future tests.  I just never got to them.

> TMK, contracts are to export only `start`.

Not necessarily, as long as there is motivation.

File: packages/run-protocol/src/feeDistributor.js
Comment (michaelfig) [reply packages/run-protocol/src/feeDistributor.js]: Agreed on constness.  I'll leave it for future `function` keyword cleanups that may be more extensive.

**Pull request merged**
